### PR TITLE
Optional onIdle(count) method added

### PR DESCRIPTION
**Pull request by treblereel**
In some cases it's very useful to be able to get a change set after some predefined period, for example, when we use APT and we want to get a list of generated files and we
 know that the whole generation stage usually takes 500ms. In this PR i introduce the simple mechanism that calls an onIdle method  after timeout expired and passes the number of events within 
that timeout.

@mdproctor ^^

### CODE REVIEW

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 1]
```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
...
+            timer.schedule(currentOnIdleTimerTask, timeout);
+          }
+          key = watchService.take();
+        }
       } catch (InterruptedException x) {
         return;
       }
       for (WatchEvent<?> event : key.pollEvents()) {
+        eventCount.incrementAndGet();
+        if (currentOnIdleTimerTask != null && !currentOnIdleTimerTask.isDone) {
```
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:263]: Does this need to be in the 'for' loop? Isn't it enough to do this  directly after the take() returns?

I don't think this is thread safe, as the timer will set isDone, which is different to the directory watcher main loop thread.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 2]
```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
...
+    final AtomicInteger eventCount = new AtomicInteger(0);
     while (listener.isWatching()) {
       // wait for key to be signalled
       WatchKey key;
       try {
-        key = watchService.take();
+        key = watchService.poll();
+        if (key == null) {
+          if (timeout > 0) {
+            currentOnIdleTimerTask = new OnIdleTimerTask(listener, eventCount.get());
```
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:253]: This creates a new OnIdleTimerTask each time. What is an OnIdleTimerTask already exists, does it need to be cancelled? Is all access got this field, and new instances to this field thread safe?

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 3]
```diff
@@ -166,6 +178,10 @@ public static Builder builder() {
   // set to null until we check if FILE_TREE is supported
   private Boolean fileTreeSupported = null;
   private FileHasher fileHasher;
+  private int timeout = -1;
+
+  private Timer timer = new Timer("DirectoryWatcherTimer");
```
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:183]: Make this final? Then it can be synchronised on.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 4]
```diff
@@ -428,4 +459,25 @@ private void notifyCreateEvent(boolean isDirectory, Path path, int count, Path r
     }
     onEvent(EventType.CREATE, isDirectory, path, count, rootPath);
   }
+
+  private class OnIdleTimerTask extends TimerTask {
+
+    private boolean isDone = false;
```
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:465]: There is unthread safe access to this.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 5]
```diff
@@ -428,4 +459,25 @@ private void notifyCreateEvent(boolean isDirectory, Path path, int count, Path r
...
+
+    private OnIdleTimerTask(DirectoryChangeListener listener, int counter) {
+      this.listener = listener;
+      this.counter = counter;
+    }
+
+    @Override
+    public void run() {
+      logger.debug("Task performed on: " + new Date() + " with event count " + counter);
+      listener.onIdle(counter);
```
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:479]: We should document that this operation is not atomic - i.e. the check on isDone() and cancel() is not atomic and may overlap with this sections onIdle() call and setting isDone. And thus it is possible that actually the timer is cancelled, while this event is being fired. I would imagine for the type of use cases this is being applied to, that is not a problem, but either way should be in in the JavaDocs.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java (hunk 3 above)
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:183]: This can be final. Further set it in the constructor. As this way you can set it null if timeout is still <= 0.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 6]
```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
...
+          }
+          key = watchService.take();
+        }
       } catch (InterruptedException x) {
         return;
       }
       for (WatchEvent<?> event : key.pollEvents()) {
+        eventCount.incrementAndGet();
+        if (currentOnIdleTimerTask != null && !currentOnIdleTimerTask.isDone) {
+          currentOnIdleTimerTask.cancel();
```
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:264]: Should this be nulled after it's cancelled?

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java (hunk 2 above)
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:253]: What if this instance already exists from the previous loop iteration? Maybe it should be cancelled and nulled first?

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java [hunk 7]
```diff
@@ -224,15 +242,28 @@ public DirectoryWatcher(
    * is closed.
    */
   public void watch() {
+    final AtomicInteger eventCount = new AtomicInteger(0);
```
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:245]: Why is this atomic? the main watch() loop  is single threaded, no other thread reads this value. It's just a value that is passed to the event handler.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java (hunk 1 above)
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:263]: This isDone() done call is not thread safe.

**Review by gmethvin — COMMENTED** [review]
We chatted about this before @mdproctor, but I'm a bit unsure of whether it's better to implement the timeout in `DirectoryWatcher` itself, or to do it in the listener. If we handle it in the listener, we'd have to provide both `onIdle` (when waiting for new events) and `onActive` (when new events are pending), but then `DirectoryWatcher` doesn't have to worry about the thread safety issues for the timer.

In other words, we'd have something like:

```java
key = watchService.poll();
if (key == null) {
  listener.onIdle();
  key = watchService.take();
  listener.onActive();
}
```

Then you can move all the complicated timer code into a separate `DirectoryChangeListener` implementation.

**Comment by gmethvin** [comment]
Also I think for the idle timeout you could use `ScheduledExecutorService`:

```java
ScheduledExecutorService scheduledExecutor = Executors.newSingleThreadScheduledExecutor();
ScheduledFuture<?> idleFuture = null;

void onIdle(int count) { // called when we start waiting for new events
  idleFuture = scheduledExecutor.schedule(() -> { onIdleTimedOut(count); }, idleTimeout, idleTimeoutUnit);
}

void onActive(int count) { // called when we start getting events again
  if (idleFuture != null) {
    idleFuture.cancel(true);
    idleFuture = null;
  }
}

void onIdleTimedOut(int count); // call this when the timeout has passed after `onIdle`.
```

Also with `ScheduledExecutorService` it'd be easier to keep track of multiple timeout tasks if you want to do that.

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java (hunk 3 above)
Comment (treblereel) [reply core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:183]: other final or constructor case, right ?

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java (hunk 7 above)
Comment (treblereel) [reply core/src/main/java/io/methvin/watcher/DirectoryWatcher.java:245]: yeap, agree

**Comment by treblereel** [comment]
I have switched to ScheduledExecutorService and transfered all OnIdle logic into ChangeSetListener, is it closer to what we want ?

As you can see, i have implemented additional ChangeSetListener constructor that accepts timeout and consumer, because current ChangeSetListener doesn't allow to subscribe to onIdle(int) directly.


wdyt ?

File: core/src/main/java/io/methvin/watcher/DirectoryWatcher.java (hunk 2 above)
Comment (treblereel) [reply core/src/main/java/io/methvin/watcher/DirectoryWatcher.java]: fixed

File: core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java [hunk 8]
```diff
@@ -44,6 +67,11 @@ public void onEvent(DirectoryChangeEvent event) {
         case OVERFLOW:
           throw new IllegalStateException("OVERFLOW not yet handled");
       }
+
+      if (timeout > 0 && currentTask != null) {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java:71]: I don't think we need `timeout > 0` since all we care about is there's a task to cancel.

File: core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java [hunk 9]
```diff
@@ -13,7 +18,25 @@
 
   private final Object lock = new Object() {};
 
-  public ChangeSetListener() {}
+  private final int timeout;
+
+  private ScheduledExecutorService service;
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java:23]: nit: can be final

File: core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java [hunk 10]
```diff
@@ -13,7 +18,25 @@
...
+
+  public ChangeSetListener() {
+    this(-1, null);
+  }
+
+  public ChangeSetListener(int timeout, Consumer<Integer> onIdleListener) {
+    this.timeout = timeout;
+    this.onIdleListener = onIdleListener;
+
+    if (timeout > 0) {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java:36]: I actually think there's a valid use case for `timeout == 0`, in case you want to be notified as soon as we hit an idle state.

**Review by gmethvin — COMMENTED** [review]
Is there a reason this needs to be in `ChangeSetListener` in particular? Do you plan to use it along with the other `ChangeSetListener` functionality?

Based on your test it seems to be entirely separate functionality from the rest of `ChangeSetListener`.

File: core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java (hunk 8 above)
Comment (treblereel) [reply core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java:71]: currentTask is null if we don't use onIdle

File: core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java (hunk 9 above)
Comment (treblereel) [reply core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java:23]: ScheduledExecutorService is setted only if we need OnIdle

File: core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java (hunk 10 above)
Comment (treblereel) [reply core/src/main/java/io/methvin/watcher/changeset/ChangeSetListener.java:36]: yeap, agree

**Comment by treblereel** [comment]
@gmethvin if i am not mistaken, Mark mentioned that this functionality can be moved from DirectoryWatcher to ChangeSetListener. We are planning to use OnIdle functionality with ChangeSetListener, but, to be honest, i can't say, i am happy with my current implementation, adding listener to listener doesn't look nature.

So if you have any ideas how to improve it i ll be happy to implement it

**Comment by gmethvin** [comment]
@mdproctor @treblereel I'm just trying to understand the overall use case so we can design the best API. This is why I'm asking how it's meant to be used. As it stands right now, it appears the idle timeout functionality could be a separate listener type rather than part of `ChangeSetListener`.

Is the idea that you want `ChangeSetListener` to retrieve a new change set _the next time the watcher is idle_? In that case, for example, you could have an API that blocks until idle and retrieves the change set of all changes since the last time the watcher was idle. On its own it just seems awkward to have the idle listener as part of `ChangeSetListener` since it doesn't seem directly related to rest of the functionality.

**Comment by mdproctor** [comment]
I think you are right. While the onIdle notification needs to be in core. We can move the timeout notification to it's own listener. We can then ask users to add the the pair of listeners (IdleTimeOutListener, ChangeSetListener), via  composite holder. Keeping the timer logic in it's own class like that, will mean easier to maintain and test code.

File: core/src/main/java/io/methvin/watcher/OnTimeoutListener.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+        if (currentTask != null) {
+          currentTask.cancel(false);
+          currentTask = null;
+        }
+        currentTask = service.schedule(() -> onTimeout(count), timeout, TimeUnit.MILLISECONDS);
+      }
+    }
+  }
+
+  public abstract void onTimeout(int count);
```
Comment (mdproctor) [inline core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:62]: do people prefer to extend this class, or pass in a lambda? A lambda would feel more modern, less lines of code.

File: core/src/main/java/io/methvin/watcher/OnTimeoutListener.java [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+      if (timeout >= 0 && currentTask != null) {
+        currentTask.cancel(false);
+        currentTask = null;
+      }
+    }
+  }
+
+  @Override
+  public void onIdle(int count) {
+    synchronized (lock) {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:51]: what do you think about making this an `AtomicReference`?

```scala
if (timeout >= 0) {
  currentTaskRef.getAndUpdate(oldTask -> {
    if (oldTask != null) {
      oldTask.cancel(false);
    }
    return service.schedule(() -> onTimeout(count), timeout, TimeUnit.MILLISECONDS);
  });
}
```

Then above you could do something like:

```java
ScheduledFuture taskToCancel = currentTaskRef.getAndSet(null);
if (taskToCancel != null) {
  taskToCancel.cancel(false);
}
```

File: core/src/main/java/io/methvin/watcher/OnTimeoutListener.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+public abstract class OnTimeoutListener implements DirectoryChangeListener {
+
+  private final Object lock = new Object() {};
+
+  private final int timeout;
+
+  private ScheduledExecutorService service;
+  private ScheduledFuture currentTask;
+
+  public OnTimeoutListener(int timeout) {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:31]: just for clarity I'd name it with the units: `timeoutMillis`. We could also have a length and a `TimeUnit` but not sure if that's overkill.

File: core/src/main/java/io/methvin/watcher/OnTimeoutListener.java (hunk 11 above)
Comment (gmethvin) [reply core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:62]: I would prefer having the user pass a lambda since it allows for a cleaner API.

File: core/src/main/java/io/methvin/watcher/OnTimeoutListener.java (hunk 13 above)
Comment (gmethvin) [reply core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:31]: Also maybe we should just throw an `IllegalArgumentException` if the timeout is negative? Is there a use case for using this listener without a timeout?

**Review by gmethvin — COMMENTED** [review]
A couple questions that came up in my review:
 1. I noticed we cancel the idle state when `onEvent` is called, but technically the watcher is no longer "idle" as soon as it starts processing a new event. There could be a delay between when the watcher receives the event and when `onEvent` fires, especially if we have hashing enabled. Does that difference matter? The solution I think would be to add a call like `onActive` at the beginning of the loop.
 2. I'm assuming most users would want to track both the idle state and the individual events, so we probably want a way to do both. One possible solution would be to have the `OnTimeoutListener` forward to another user-defined `DirectoryChangeListener`. Or we could have a way to create a composite listener from a list of listeners.

**Comment by gmethvin** [comment]
@mdproctor I think either way there's a use case for a helper like `DirectoryChangeListener.fromListeners(listeners)`.

In this case though I was wondering if we should have something like `DirectoryChangeListener.withIdleDelay(timeoutMillis, listener)`, which would just delay the call to `listener.onIdle` with your timeout code. Not totally set on that API though.

File: core/src/main/java/io/methvin/watcher/OnTimeoutListener.java (hunk 13 above)
Comment (treblereel) [reply core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:31]: @gmethvin nope, i am pretty sure nope

File: core/src/main/java/io/methvin/watcher/OnTimeoutListener.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+                  }
+                },
+                timeoutMillis,
+                TimeUnit.MILLISECONDS);
+          });
+    }
+  }
+
+  public OnTimeoutListener onTimeout(Consumer<Integer> consumer) {
+    this.consumer = consumer;
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:68]: I'd prefer to keep the API immutable if possible, and as-is this API doesn't make it obvious that this will replace the existing consumer.

I would suggest we pass a single `Consumer onIdleTimeout` in the constructor, or in a builder.

File: core/src/main/java/io/methvin/watcher/OnTimeoutListener.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+package io.methvin.watcher;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+public class OnTimeoutListener implements DirectoryChangeListener {
```
Comment (gmethvin) [inline core/src/main/java/io/methvin/watcher/OnTimeoutListener.java:24]: Can this be final?

File: core/src/test/java/io/methvin/watcher/DirectoryWatcherOnIdleTest.java [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+    assertEquals(1, onIdleCalled.get());
+
+    Map<Path, ChangeSet> changeSet = composite.getChangeSet();
+    assertTrue(!changeSet.isEmpty());
+    assertEquals(7, changeSet.get(d1).created().size());
+
+    watcher.close();
+  }
+
+  private class CompositeListener implements DirectoryChangeListener {
```
Comment (gmethvin) [inline core/src/test/java/io/methvin/watcher/DirectoryWatcherOnIdleTest.java:190]: I thought we were talking about adding a helper that creates a listener out of a list of listeners? Seems like a natural addition to the API.

**Comment by gmethvin** [comment]
@mdproctor @treblereel I merged this along with `05c155e02871a7b6f6c73a11d58b001fd9f38ed7` which addresses my comments. Let me know if that works.
